apps/prairielearn/src/ee/pages/instructorAiGenerateQuestion/instructorAiGenerateQuestion.ts

```diff
@@ -181,29 +225,59 @@ router.post(
         );
       }
     } else if (req.body.__action === 'save_question') {
-      const genJobs = await selectJobsByJobSequenceId(req.body.unsafe_sequence_job_id);
-      if (
-        genJobs.length !== 1 ||
-        !genJobs[0]?.course_id ||
-        !idsEqual(genJobs[0]?.course_id, res.locals.course.id)
-      ) {
-        throw new error.HttpStatusError(
-          403,
-          `Job sequence ${req.body.unsafe_sequence_job_id} not found.`,
-        );
+      const draftQid = `__drafts__/${req.body.unsafe_qid}`;
+      const threads = await queryRows(
+        sql.select_generation_thread_items,
+        { qid: draftQid, course_id: res.locals.course.id.toString() },
+        AiGenerationPromptSchema,
+      );
+
+      if (threads.length === 0) {
+        throw new error.HttpStatusError(403, `Draft question ${req.body.unsafe_qid} not found.`);
       }
 
       const qid = await saveGeneratedQuestion(
         res,
-        genJobs[0]?.data?.html,
-        genJobs[0]?.data?.python,
+        threads[threads.length - 1].html || undefined,
+        threads[threads.length - 1].python || undefined,
+        req.body.title,
+        req.body.qid,
       );
 
       res.redirect(res.locals.urlPrefix + '/question/' + qid + '/settings');
+    } else if (req.body.__action === 'grade' || req.body.__action === 'save') {
+      res.locals.question = await queryRow(
+        sql.select_question_by_qid_and_course,
+        { qid: `__drafts__/${req.query.qid}`, course_id: res.locals.course.id },
+        QuestionSchema,
+      );
+      const variantId = await processSubmission(req, res);
+      res.redirect(
+        `${res.locals.urlPrefix}/ai_generate_question?qid=${req.query?.qid}&variant_id=${variantId}`,
+      );
     } else {
       throw new error.HttpStatusError(400, `Unknown action: ${req.body.__action}`);
     }
   }),
 );
 
+router.get(
+  '/variant/:variant_id(\\d+)/submission/:submission_id(\\d+)',
```

Comment:
this partial renders submission results (from backend grading).

Reply:
Turns out this isn't necessary, it'll use the route from the instructor question preview, which will always also work in this context.